shared/engine/index.js

```diff
@@ -1,435 +1,290 @@
-          }
-          return
-        }
-        once = true
-
-        if (printRPC) {
-          this.logLocal('RPC ▼ error: ', method, param, ...args)
-        }
+    console.warn(`Unknown incoming rpc: ${sessionID} ${method}`)
 
-        if (response) {
-          cleanup()
-          response.error(...args)
-        } else if (__DEV__) {
-          console.warn('Calling response.error on non-response object: ', method)
-        }
-      },
+    if (__DEV__ && this._failOnError) {
+      throw new Error(`unhandled incoming rpc: ${sessionID} ${method} ${JSON.stringify(param)}${response ? '. has response' : ''}`)
     }
 
-    if (__DEV__ && KEYBASE_RPC_DELAY) {
-      const result = wrappedResponse.result
-      const error = wrappedResponse.error
-      wrappedResponse.result = (...args) => {
-        console.log('RPC Delayed')
-        setTimeout(() => { result(...args) }, KEYBASE_RPC_DELAY)
-      }
-      wrappedResponse.error = (...args) => {
-        console.log('RPC Delayed')
-        setTimeout(() => { error(...args) }, KEYBASE_RPC_DELAY)
-      }
-    }
+    response && response.error && response.error({
+      code: ConstantsStatusCode.scgeneric,
+      desc: `Unhandled incoming RPC ${sessionID} ${method}`,
+    })
+  }
 
-    // Incoming calls have no sessionID so let's ignore
-    if (sessionID) {
-      this.sessionIDToResponse[String(sessionID)] = wrappedResponse
+  // An incoming rpc call
+  _rpcIncoming (payload: {method: MethodKey, param: Array<Object>, response: ?Object}) {
+    const {method, param: incomingParam, response} = payload
+    const param = incomingParam && incomingParam.length ? incomingParam[0] : {}
+    const {seqid, cancelled} = response || {seqid: 0, cancelled: false}
+    const {sessionID} = param
 
-      if (printOutstandingRPCs) {
-        this.responseMeta[sessionID] = {method, param}
+    if (cancelled) {
+      this._handleCancel(seqid)
+    } else {
+      const session = this._sessionsMap[String(sessionID)]
+      if (session && session.incomingCall(method, param, response)) { // Part of a session?
+      } else if (this._incomingHandler[method]) { // General incoming
+        const handler = this._incomingHandler[method]
+        rpcLog('engineInternal', 'Handling incoming')
+        handler(param, response)
+      } else { // Unhandled
+        this._handleUnhandled(sessionID, method, seqid, param, response)
       }
     }
-
-    if (seqid) {
-      this.seqIDToWrappedResponse[seqid] = wrappedResponse
-    }
-
-    return wrappedResponse
   }
 
-  _generalIncomingRpc (method: MethodKey, param: Object, response: any) {
-    const listener = this.generalListeners[method]
+  // An outgoing call. ONLY called by the flow-type rpc helpers
+  _rpcOutgoing (params: {
+    method: MethodKey,
+    param?: ?Object,
+    incomingCallMap?: incomingCallMapType,
+    callback?: ?(...args: Array<any>) => void,
+    waitingHandler?: WaitingHandlerType}) {
```

Comment:
It's pretty hard to tell where the arguments end and the functions begin with this style. Maybe drop the closing `) {` to the next line and dedent?